// @vocalbridgeai/sdk/dist/index.js
'use strict';

var livekitClient = require('livekit-client');

// src/VocalBridge.ts

// src/types.ts
var ConnectionState = /* @__PURE__ */ ((ConnectionState2) => {
  ConnectionState2["Disconnected"] = "disconnected";
  ConnectionState2["Connecting"] = "connecting";
  ConnectionState2["WaitingForAgent"] = "waiting_for_agent";
  ConnectionState2["Connected"] = "connected";
  ConnectionState2["Reconnecting"] = "reconnecting";
  ConnectionState2["Disconnecting"] = "disconnecting";
  return ConnectionState2;
})(ConnectionState || {});

// src/errors.ts
var VocalBridgeErrorCode = /* @__PURE__ */ ((VocalBridgeErrorCode2) => {
  VocalBridgeErrorCode2["TokenFetchFailed"] = "TOKEN_FETCH_FAILED";
  VocalBridgeErrorCode2["ConnectionFailed"] = "CONNECTION_FAILED";
  VocalBridgeErrorCode2["MicrophoneError"] = "MICROPHONE_ERROR";
  VocalBridgeErrorCode2["DataChannelError"] = "DATA_CHANNEL_ERROR";
  VocalBridgeErrorCode2["ReconnectFailed"] = "RECONNECT_FAILED";
  VocalBridgeErrorCode2["UsageLimitExceeded"] = "USAGE_LIMIT_EXCEEDED";
  VocalBridgeErrorCode2["AgentNotFound"] = "AGENT_NOT_FOUND";
  VocalBridgeErrorCode2["AgentNotActive"] = "AGENT_NOT_ACTIVE";
  VocalBridgeErrorCode2["InvalidConfig"] = "INVALID_CONFIG";
  return VocalBridgeErrorCode2;
})(VocalBridgeErrorCode || {});
var VocalBridgeError = class extends Error {
  constructor(message, code, cause) {
    super(message);
    this.name = "VocalBridgeError";
    this.code = code;
    this.cause = cause;
  }
};

// src/events.ts
var TypedEventEmitter = class {
  constructor() {
    this.listeners = /* @__PURE__ */ new Map();
  }
  on(event, listener) {
    if (!this.listeners.has(event)) {
      this.listeners.set(event, /* @__PURE__ */ new Set());
    }
    this.listeners.get(event).add(listener);
    return this;
  }
  off(event, listener) {
    this.listeners.get(event)?.delete(listener);
    return this;
  }
  once(event, listener) {
    const wrapper = ((data) => {
      this.off(event, wrapper);
      listener(data);
    });
    return this.on(event, wrapper);
  }
  emit(event, data) {
    const set = this.listeners.get(event);
    if (set) {
      for (const listener of set) {
        try {
          listener(data);
        } catch {
        }
      }
    }
  }
  removeAllListeners() {
    this.listeners.clear();
  }
};

// src/constants.ts
var DATA_CHANNEL_TOPIC = "client_actions";
var DEFAULT_API_BASE_URL = "https://vocalbridgeai.com";
var PROTOCOL_ACTIONS = {
  HEARTBEAT: "heartbeat",
  HEARTBEAT_ACK: "heartbeat_ack",
  SEND_TRANSCRIPT: "send_transcript",
  QUERY_AGENT: "query_agent",
  AGENT_RESPONSE: "agent_response"
};

// src/auth.ts
var TOKEN_FETCH_TIMEOUT_MS = 15e3;
function isApiKeyAuth(auth) {
  return "apiKey" in auth;
}
function isTokenUrlAuth(auth) {
  return "tokenUrl" in auth;
}
function isTokenProviderAuth(auth) {
  return "tokenProvider" in auth;
}
function createTokenFetcher(config, participantName, sessionId) {
  if (isTokenProviderAuth(config)) {
    return async () => {
      try {
        return await config.tokenProvider();
      } catch (err) {
        throw new VocalBridgeError(
          "Token provider failed",
          "TOKEN_FETCH_FAILED" /* TokenFetchFailed */,
          err
        );
      }
    };
  }
  if (isTokenUrlAuth(config)) {
    return async () => {
      const body = {
        participant_name: participantName,
        ...config.body
      };
      if (sessionId) body.session_id = sessionId;
      try {
        const controller = new AbortController();
        const timeout = setTimeout(
          () => controller.abort(),
          TOKEN_FETCH_TIMEOUT_MS
        );
        const response = await fetch(config.tokenUrl, {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            ...config.headers
          },
          body: JSON.stringify(body),
          signal: controller.signal
        });
        clearTimeout(timeout);
        if (!response.ok) {
          throw await mapTokenError(response);
        }
        return await response.json();
      } catch (err) {
        if (err instanceof VocalBridgeError) throw err;
        throw new VocalBridgeError(
          "Failed to fetch token from URL",
          "TOKEN_FETCH_FAILED" /* TokenFetchFailed */,
          err
        );
      }
    };
  }
  if (isApiKeyAuth(config)) {
    if (typeof globalThis.window !== "undefined") {
      console.warn(
        "[VocalBridge] API key auth detected in a browser environment. This exposes your API key to end users. Use tokenUrl or tokenProvider for production."
      );
    }
    return async () => {
      const headers = {
        "Content-Type": "application/json",
        "X-API-Key": config.apiKey
      };
      if (config.agentId) {
        headers["X-Agent-Id"] = config.agentId;
      }
      const body = {
        participant_name: participantName
      };
      if (sessionId) body.session_id = sessionId;
      try {
        const controller = new AbortController();
        const timeout = setTimeout(
          () => controller.abort(),
          TOKEN_FETCH_TIMEOUT_MS
        );
        const response = await fetch(
          `${DEFAULT_API_BASE_URL}/api/v1/token`,
          {
            method: "POST",
            headers,
            body: JSON.stringify(body),
            signal: controller.signal
          }
        );
        clearTimeout(timeout);
        if (!response.ok) {
          throw await mapTokenError(response);
        }
        return await response.json();
      } catch (err) {
        if (err instanceof VocalBridgeError) throw err;
        throw new VocalBridgeError(
          "Failed to fetch token with API key",
          "TOKEN_FETCH_FAILED" /* TokenFetchFailed */,
          err
        );
      }
    };
  }
  throw new VocalBridgeError(
    "Invalid auth configuration: provide apiKey, tokenUrl, or tokenProvider",
    "INVALID_CONFIG" /* InvalidConfig */
  );
}
async function mapTokenError(response) {
  let detail = "";
  try {
    const body = await response.json();
    detail = body.detail || body.message || "";
  } catch {
  }
  switch (response.status) {
    case 401:
      return new VocalBridgeError(
        detail || "Invalid API key",
        "TOKEN_FETCH_FAILED" /* TokenFetchFailed */
      );
    case 403:
      return new VocalBridgeError(
        detail || "Usage limit exceeded",
        "USAGE_LIMIT_EXCEEDED" /* UsageLimitExceeded */
      );
    case 404:
      return new VocalBridgeError(
        detail || "Agent not found",
        "AGENT_NOT_FOUND" /* AgentNotFound */
      );
    default:
      return new VocalBridgeError(
        detail || `Token request failed (${response.status})`,
        "TOKEN_FETCH_FAILED" /* TokenFetchFailed */
      );
  }
}
var AudioManager = class {
  constructor(room, options) {
    this.audioElements = /* @__PURE__ */ new Map();
    this.micEnabled = false;
    this.handleTrackSubscribed = (track, publication, participant) => {
      if (track.kind === livekitClient.Track.Kind.Audio) {
        const audioElement = track.attach();
        audioElement.style.display = "none";
        document.body.appendChild(audioElement);
        this.audioElements.set(track.sid, audioElement);
      }
    };
    this.handleTrackUnsubscribed = (track) => {
      if (track.kind === livekitClient.Track.Kind.Audio && track.sid) {
        const el = this.audioElements.get(track.sid);
        if (el) {
          el.srcObject = null;
          el.remove();
          this.audioElements.delete(track.sid);
        }
        track.detach().forEach((el2) => el2.remove());
      }
    };
    this.room = room;
    this.autoPlay = options.autoPlay;
    if (this.autoPlay) {
      this.room.on(
        livekitClient.RoomEvent.TrackSubscribed,
        this.handleTrackSubscribed
      );
      this.room.on(
        livekitClient.RoomEvent.TrackUnsubscribed,
        this.handleTrackUnsubscribed
      );
    }
  }
  /** Start capturing from microphone and publish to room. */
  async startMicrophone(captureDefaults) {
    await this.room.localParticipant.setMicrophoneEnabled(true, {
      echoCancellation: captureDefaults?.echoCancellation ?? true,
      noiseSuppression: captureDefaults?.noiseSuppression ?? true,
      autoGainControl: captureDefaults?.autoGainControl ?? true
    });
    this.micEnabled = true;
  }
  /** Set microphone enabled/disabled. */
  async setEnabled(enabled) {
    await this.room.localParticipant.setMicrophoneEnabled(enabled);
    this.micEnabled = enabled;
  }
  /** Whether mic is currently enabled. */
  get isEnabled() {
    return this.micEnabled;
  }
  /** Clean up all audio elements and remove event listeners. */
  dispose() {
    for (const [sid, el] of this.audioElements) {
      el.srcObject = null;
      el.remove();
      this.audioElements.delete(sid);
    }
    if (this.autoPlay) {
      this.room.off(
        livekitClient.RoomEvent.TrackSubscribed,
        this.handleTrackSubscribed
      );
      this.room.off(
        livekitClient.RoomEvent.TrackUnsubscribed,
        this.handleTrackUnsubscribed
      );
    }
  }
};

// src/transcript.ts
var DEFAULT_MAX_ENTRIES = 5e3;
var TranscriptAccumulator = class {
  constructor(maxEntries = DEFAULT_MAX_ENTRIES) {
    this.entries = [];
    this.maxEntries = maxEntries;
  }
  /** Add a new transcript entry. Evicts oldest entries if limit is exceeded. */
  add(entry) {
    this.entries.push(entry);
    if (this.entries.length > this.maxEntries) {
      this.entries = this.entries.slice(-this.maxEntries);
    }
  }
  /** Get all transcript entries (read-only). */
  getAll() {
    return this.entries;
  }
  /** Clear all transcript entries. */
  clear() {
    this.entries = [];
  }
};

// src/data-channel.ts
var encoder = new TextEncoder();
var decoder = new TextDecoder();
function encodeAction(action, payload = {}) {
  const message = {
    type: "client_action",
    action,
    payload
  };
  return encoder.encode(JSON.stringify(message));
}
function decodeMessage(data) {
  try {
    const text = decoder.decode(data);
    const parsed = JSON.parse(text);
    if (parsed && typeof parsed === "object" && parsed.type === "client_action" && typeof parsed.action === "string" && parsed.action.length <= 1024) {
      const payload = parsed.payload;
      const safePayload = typeof payload === "object" && payload !== null && !Array.isArray(payload) ? payload : {};
      return {
        type: "client_action",
        action: parsed.action,
        payload: safePayload
      };
    }
    return null;
  } catch {
    return null;
  }
}

// src/VocalBridge.ts
var VocalBridge = class extends TypedEventEmitter {
  constructor(options) {
    super();
    this._room = null;
    this.audioManager = null;
    this.transcriptAccumulator = new TranscriptAccumulator();
    this._state = "disconnected" /* Disconnected */;
    this._isMicrophoneEnabled = false;
    this.reconnectAttempts = 0;
    this._pendingOperation = null;
    this.aiAgentQueryHandler = null;
    this.options = {
      participantName: "User",
      autoAckHeartbeat: true,
      autoPlayAudio: true,
      maxReconnectAttempts: 3,
      reconnectDelay: 2e3,
      debug: false,
      ...options
    };
    this.fetchToken = createTokenFetcher(
      this.options.auth,
      this.options.participantName,
      this.options.sessionId
    );
  }
  // ---- Lifecycle ----
  /** Connect to the voice agent. Fetches token, establishes connection, publishes mic. */
  async connect() {
    if (this._pendingOperation) {
      await this._pendingOperation;
    }
    if (this._state !== "disconnected" /* Disconnected */ && this._state !== "reconnecting" /* Reconnecting */) {
      this.log("Already connected or connecting");
      return;
    }
    const op = this._connect();
    this._pendingOperation = op;
    try {
      await op;
    } finally {
      if (this._pendingOperation === op) {
        this._pendingOperation = null;
      }
    }
  }
  async _connect() {
    this.setState("connecting" /* Connecting */);
    try {
      this.log("Fetching token...");
      const tokenResponse = await this.fetchToken();
      this._agentMode = tokenResponse.agent_mode;
      this._roomName = tokenResponse.room_name;
      this._participantIdentity = tokenResponse.participant_identity;
      this._room = new livekitClient.Room({
        adaptiveStream: true,
        dynacast: true,
        audioCaptureDefaults: {
          echoCancellation: this.options.audioCaptureDefaults?.echoCancellation ?? true,
          noiseSuppression: this.options.audioCaptureDefaults?.noiseSuppression ?? true,
          autoGainControl: this.options.audioCaptureDefaults?.autoGainControl ?? true
        }
      });
      this.setupRoomListeners(this._room);
      this.audioManager = new AudioManager(this._room, {
        autoPlay: this.options.autoPlayAudio
      });
      this.log("Connecting...");
      const connectionUrl = tokenResponse.url || tokenResponse.livekit_url;
      if (!connectionUrl) {
        throw new VocalBridgeError(
          "Token response missing connection URL",
          "TOKEN_FETCH_FAILED" /* TokenFetchFailed */
        );
      }
      await this._room.connect(connectionUrl, tokenResponse.token);
      this.setState("waiting_for_agent" /* WaitingForAgent */);
      for (const participant of this._room.remoteParticipants.values()) {
        if (this.isAgentParticipant(participant)) {
          this.setState("connected" /* Connected */);
          break;
        }
      }
      this.log("Starting microphone...");
      await this.audioManager.startMicrophone(
        this.options.audioCaptureDefaults
      );
      this._isMicrophoneEnabled = true;
      this.emit("microphoneChanged", true);
      this.reconnectAttempts = 0;
      this.log("Connected successfully");
    } catch (err) {
      this.setState("disconnected" /* Disconnected */);
      const error = err instanceof VocalBridgeError ? err : new VocalBridgeError(
        "Connection failed",
        "CONNECTION_FAILED" /* ConnectionFailed */,
        err
      );
      this.emit("error", error);
      throw error;
    }
  }
  /** Disconnect from the voice agent. Cleans up all resources. */
  async disconnect() {
    if (this._pendingOperation) {
      await this._pendingOperation;
    }
    if (this._state === "disconnected" /* Disconnected */) return;
    const op = this._disconnect();
    this._pendingOperation = op;
    try {
      await op;
    } finally {
      if (this._pendingOperation === op) {
        this._pendingOperation = null;
      }
    }
  }
  async _disconnect() {
    this.setState("disconnecting" /* Disconnecting */);
    try {
      this.audioManager?.dispose();
      this.audioManager = null;
      if (this._room) {
        await this._room.disconnect();
        this._room = null;
      }
    } finally {
      this._isMicrophoneEnabled = false;
      this.setState("disconnected" /* Disconnected */);
    }
  }
  // ---- Audio Controls ----
  /** Mute or unmute the microphone. */
  async setMicrophoneEnabled(enabled) {
    if (!this.audioManager) {
      throw new VocalBridgeError(
        "Not connected",
        "CONNECTION_FAILED" /* ConnectionFailed */
      );
    }
    await this.audioManager.setEnabled(enabled);
    this._isMicrophoneEnabled = enabled;
    this.emit("microphoneChanged", enabled);
  }
  /** Toggle microphone mute state. */
  async toggleMicrophone() {
    await this.setMicrophoneEnabled(!this._isMicrophoneEnabled);
  }
  // ---- Client Actions ----
  /** Send a custom action to the agent via the data channel. */
  async sendAction(action, payload = {}) {
    if (!this._room?.localParticipant) {
      throw new VocalBridgeError(
        "Not connected \u2014 cannot send action",
        "DATA_CHANNEL_ERROR" /* DataChannelError */
      );
    }
    const data = encodeAction(action, payload);
    await this._room.localParticipant.publishData(data, {
      reliable: true,
      topic: DATA_CHANNEL_TOPIC
    });
  }
  /** Send a response to an AI agent query. */
  async sendAIAgentResponse(turnId, response) {
    await this.sendAction(PROTOCOL_ACTIONS.AGENT_RESPONSE, {
      response: response.slice(0, 5e4),
      turn_id: turnId
    });
  }
  // ---- AI Agent Query Handler ----
  /**
   * Register a handler for AI Agent queries.
   * The return value is automatically sent back as agent_response.
   * Returns an unsubscribe function.
   */
  onAIAgentQuery(handler) {
    this.aiAgentQueryHandler = handler;
    return () => {
      if (this.aiAgentQueryHandler === handler) {
        this.aiAgentQueryHandler = null;
      }
    };
  }
  // ---- Read-only State ----
  /** Current connection state. */
  get state() {
    return this._state;
  }
  /** Whether the microphone is enabled. */
  get isMicrophoneEnabled() {
    return this._isMicrophoneEnabled;
  }
  /** Accumulated transcript entries. */
  get transcript() {
    return this.transcriptAccumulator.getAll();
  }
  /** Clear all accumulated transcript entries. */
  clearTranscript() {
    this.transcriptAccumulator.clear();
  }
  /** Agent mode from the token response. */
  get agentMode() {
    return this._agentMode;
  }
  /** Room name from the current session. */
  get roomName() {
    return this._roomName;
  }
  /** Participant identity from the current session. */
  get participantIdentity() {
    return this._participantIdentity;
  }
  /**
   * Access the underlying Room for advanced use cases.
   * Using this directly bypasses SDK validation and size limits.
   */
  get room() {
    return this._room;
  }
  // ---- Private ----
  setState(state) {
    if (this._state === state) return;
    this._state = state;
    this.log(`State: ${state}`);
    this.emit("connectionStateChanged", state);
  }
  setupRoomListeners(room) {
    room.on(
      livekitClient.RoomEvent.ParticipantConnected,
      (participant) => {
        this.log(`Participant connected: ${participant.identity}`);
        if (this._state === "waiting_for_agent" /* WaitingForAgent */ && this.isAgentParticipant(participant)) {
          this.setState("connected" /* Connected */);
        }
      }
    );
    room.on(
      livekitClient.RoomEvent.TrackSubscribed,
      (track, publication, participant) => {
        if (track.kind === livekitClient.Track.Kind.Audio && this._state === "waiting_for_agent" /* WaitingForAgent */) {
          this.setState("connected" /* Connected */);
        }
      }
    );
    room.on(livekitClient.RoomEvent.Disconnected, () => {
      this.log("Room disconnected");
      if (this._state !== "disconnecting" /* Disconnecting */) {
        this.handleUnexpectedDisconnect();
      }
    });
    room.on(livekitClient.RoomEvent.DataReceived, (payload, participant, kind, topic) => {
      if (topic === DATA_CHANNEL_TOPIC) {
        this.handleDataChannelMessage(payload);
      }
    });
    room.on(livekitClient.RoomEvent.MediaDevicesError, (error) => {
      this.emit(
        "error",
        new VocalBridgeError(
          error?.message || "Microphone access denied",
          "MICROPHONE_ERROR" /* MicrophoneError */,
          error
        )
      );
    });
  }
  handleDataChannelMessage(payload) {
    const message = decodeMessage(payload);
    if (!message) return;
    switch (message.action) {
      case PROTOCOL_ACTIONS.HEARTBEAT: {
        const heartbeat = {
          timestamp: typeof message.payload.timestamp === "number" ? message.payload.timestamp : Date.now(),
          agent_identity: typeof message.payload.agent_identity === "string" ? message.payload.agent_identity : "unknown"
        };
        this.emit("heartbeat", heartbeat);
        if (this.options.autoAckHeartbeat) {
          this.sendAction(PROTOCOL_ACTIONS.HEARTBEAT_ACK, {
            timestamp: heartbeat.timestamp
          }).catch(() => {
          });
        }
        break;
      }
      case PROTOCOL_ACTIONS.SEND_TRANSCRIPT: {
        const { role, text, timestamp } = message.payload;
        if (role !== "user" && role !== "agent" || typeof text !== "string") {
          break;
        }
        const entry = {
          role,
          text: text.slice(0, 5e4),
          // Cap at 50KB to prevent memory abuse
          timestamp: typeof timestamp === "number" ? timestamp : Date.now()
        };
        this.transcriptAccumulator.add(entry);
        this.emit("transcript", entry);
        break;
      }
      case PROTOCOL_ACTIONS.QUERY_AGENT: {
        const rawQuery = message.payload.query;
        const turnId = message.payload.turn_id;
        if (typeof rawQuery !== "string" || typeof turnId !== "string") {
          break;
        }
        const query = rawQuery.slice(0, 5e4);
        if (this.aiAgentQueryHandler) {
          this.aiAgentQueryHandler(query).then((response) => {
            return this.sendAIAgentResponse(turnId, response);
          }).catch((err) => {
            this.emit(
              "error",
              new VocalBridgeError(
                "AI Agent query handler failed",
                "DATA_CHANNEL_ERROR" /* DataChannelError */,
                err
              )
            );
          });
        }
        this.emit("aiAgentQuery", { query, turnId });
        break;
      }
      case PROTOCOL_ACTIONS.HEARTBEAT_ACK:
      case PROTOCOL_ACTIONS.AGENT_RESPONSE:
        break;
      default:
        this.emit("agentAction", {
          action: message.action,
          payload: message.payload
        });
        break;
    }
  }
  isAgentParticipant(participant) {
    const id = participant.identity;
    return id !== this._participantIdentity && (id.startsWith("agent-") || id.startsWith("vb-agent-"));
  }
  async handleUnexpectedDisconnect() {
    if (this.reconnectAttempts >= this.options.maxReconnectAttempts) {
      this.audioManager?.dispose();
      this.audioManager = null;
      this._room = null;
      this._isMicrophoneEnabled = false;
      this.setState("disconnected" /* Disconnected */);
      this.emit(
        "error",
        new VocalBridgeError(
          "Reconnection failed after max attempts",
          "RECONNECT_FAILED" /* ReconnectFailed */
        )
      );
      return;
    }
    this.setState("reconnecting" /* Reconnecting */);
    this.reconnectAttempts++;
    this.log(
      `Reconnecting (attempt ${this.reconnectAttempts}/${this.options.maxReconnectAttempts})...`
    );
    this.audioManager?.dispose();
    this.audioManager = null;
    this._room = null;
    await new Promise(
      (resolve) => setTimeout(resolve, this.options.reconnectDelay)
    );
    try {
      await this.connect();
    } catch {
    }
  }
  log(message) {
    if (this.options.debug) {
      console.log(`[VocalBridge] ${message}`);
    }
  }
};

exports.ConnectionState = ConnectionState;
exports.DATA_CHANNEL_TOPIC = DATA_CHANNEL_TOPIC;
exports.PROTOCOL_ACTIONS = PROTOCOL_ACTIONS;
exports.VocalBridge = VocalBridge;
exports.VocalBridgeError = VocalBridgeError;
exports.VocalBridgeErrorCode = VocalBridgeErrorCode;
exports.decodeMessage = decodeMessage;
exports.encodeAction = encodeAction;
